libs/core/src/main/java/org/opensearch/core/common/io/stream/BytesWriteable.java

```diff
@@ -0,0 +1,70 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+/*
+ * Modifications Copyright OpenSearch Contributors. See
+ * GitHub history for details.
+ */
+
+package org.opensearch.core.common.io.stream;
+
+import org.opensearch.common.annotation.ExperimentalApi;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+
+/**
+ * Implementers can be written to a {@linkplain OutputStream} and read from a byte array. This allows them to be "thrown
+ * across the wire" using OpenSearch's internal protocol with protobuf bytes.
+ *
+ * @opensearch.api
+ */
+@ExperimentalApi
+public interface BytesWriteable {
```

Comment:
Could we generalize of the source (stream) here  instead of introducing a parallel hierarchy? 

```java
public interface StreamWriteable {
    interface Writer<O, V> {
        void write(final O out, V value) throws IOException;
    }

    interface Reader<I, V> {
        V read(final I in) throws IOException;
    }
}
```

In this case we could make the current `Writeable` and new one the part of the same hierarchy?

Reply:
I had tried generalizing the hierarchy, the problem with that is: `TransportMessage` cannot implement both `Writeable` and `BytesWriteable` as it complains that it is implementing the same thing twice from the same hierarchy. We want the message to implement both since for now both ways exist side by side. I tried multiple ways to do it but ended up with the same fix of keeping it separate for now.

Reply:
Hm ... we should be able to implement both: the classes have different input arguments, in any case - we don't have to implement both on the same class, the intermediary layers could do that.

Reply:
Ah I think I tried a little differently than mentioned here, let me add this change to a newer commit as well.